desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+
+    a, b, c, d = coef
+    d = jnp.squeeze((d[jnp.newaxis].T - constant).T)
+    t_0 = b**2 - 3 * a * c
+    t_1 = 2 * b**3 - 9 * a * b * c + 27 * a**2 * d
+    C = ((t_1 + complex_sqrt(t_1**2 - 4 * t_0**3)) / 2) ** (1 / 3)
+    C_is_zero = jnp.isclose(C, 0)
+
+    def compute_roots(xi_k):
+        t_3 = jnp.where(C_is_zero, 0, t_0 / (xi_k * C))
+        r = -(b + xi_k * C + t_3) / (3 * a)
+        return r
+
+    def clip_roots(r):
+        r = jnp.where(jnp.isreal(r), jnp.clip(jnp.real(r), a_min, a_max), a_min)
+        return r
+
+    xi_1 = (-1 + (-3) ** 0.5) / 2
+    xi_2 = xi_1**2
+    xi_3 = 1
+    roots = tuple(map(compute_roots, (xi_1, xi_2, xi_3)))
+    if clip:
+        roots = tuple(map(clip_roots, roots))
+    roots = jnp.stack(roots, axis=-1)
+    if sort:
+        roots = jnp.sort(roots, axis=-1)
+    return roots
+
+
+def _get_bounce_points(pitch, zeta, poly_B, poly_B_z):
+    """Get the bounce points given |B| and 1 / λ.
+
+    Parameters
+    ----------
+    pitch : ndarray
+        λ values representing the constant function 1 / λ.
+    zeta : ndarray
+        Field line-following ζ coordinates of spline knots.
+    poly_B : ndarray
+        Polynomial coefficients of the cubic spline of |B|.
+    poly_B_z : ndarray
+        Polynomial coefficients of the cubic spline of ∂|B|/∂_ζ.
+
+    Returns
+    -------
+    intersect, is_bp, bp1, bp2 : ndarray, ndarray, ndarray, ndarray
+        The polynomials' intersection points with 1 / λ is given by ``intersect``.
+        In order to be JIT compilable, the returned array must have a shape that
+        accommodates the case where each cubic polynomial intersects 1 / λ thrice.
+        The returned array ``intersect`` has shape (pitch.size, M * L, N, NUM_ROOTS).
+
+        The boolean mask ``is_bp`` encodes whether a given entry in ``intersect``
+        is a valid bounce point. The boolean masks ``bp1`` and ``bp2`` encode whether
+        a given entry in ``intersect`` is a valid starting and ending bounce point,
+        respectively. These arrays have shape (pitch.size * M * L, N * NUM_ROOTS).
+
+    """
+    ML = poly_B.shape[1]
+    N = poly_B.shape[2]
+    NUM_ROOTS = 3
+    a_min = zeta[:-1]
+    a_max = zeta[1:]
+    intersect = cubic_poly_roots(poly_B, 1 / pitch, a_min, a_max, sort=True).reshape(
+        pitch.size, ML, N, NUM_ROOTS
+    )
+    # Typically, this is the shape that is needed later, (as is the case in
+    # bounce_integrals because jax.vmap works over one axis at a time), so we
+    # reshape early here to prevent unnecessary data movement.
+    B_z = polyval(intersect, poly_B_z[:, jnp.newaxis]).reshape(
+        pitch.size * ML, N * NUM_ROOTS
+    )
+    # Check sign of derivative to determine whether root is a valid bounce point.
+    bp1 = B_z <= 0
+    bp2 = B_z >= 0
+    # Periodic boundary to compute bounce integrals of particles trapped outside
+    # this snapshot of the field lines.
+    is_bp = bp1 & jnp.roll(bp2, -1, axis=-1)
```

Comment:
I made a logic error here, thankfully it is isolated to these few lines. This line assumes that `intersect` is an array of the intersection points of the spline with 1/lambda along a field line. That is almost what `intersect` is. For the reason discussed in the docstring, intersect may have extraneous elements in it that are not intersects. These extranous elements were clipped to the left knot of the spline,(which is a fine trick to compute the bounce integrals as is done in the `direct` method).

So we need to first compute a mask for only the true intersections in `intersect`, then compute B_z on the elements in that mask, then we can compute is_bp as above.